### DISCUSSION TITLE
Persistent in Phalcon 4

### DISCUSSION THREAD

**Discussion by Alistar84** [phalcon/cphalcon#15523] [Q&A] [answered]
Hi,
I am upgrading from phalcon 3 to 4 and I have a problem with the persistent data.

Compared to v3 I had to insert a shared service sersessionbag because otherwise I have a PHP error (Service 'sessionBag' wasn't found in the dependency injection container).

The problem is that in v3 `$ this->persistent->parameters` saves in the same variable different data depending on which controller it is called on, but now it always shows the same value. It does not matter from which class it is called.

Checking the session, in v3 I have a key for each controller, while in v4 I only have one key called "bag" (the name of the sessionbag).

How can I solve the problem?

**Comment by quasipickle** [comment]
I encountered this exact same problem.  Unfortunately, I wasn't able to find a solution using `persistent`.  (Fortunately for me it turns out I didn't need to use `persistent` at all).

Before I came to that realization though, I was namespacing all my parameters.

So rather than storing `$this->persistent->primaryFilter` in my `UsersController`, I would store (and access) `$this->persistent->usersPrimaryFilter`.  That ensured no collisions.

**Reply by niden** [reply]
Actually this is a very good technique to ensure no collisions are happening. 

Curious though why the bag caused all these issues. To my knowledge the SessionBag was one of the few components that was not touched from v3 to v4

**Reply by quasipickle** [reply]
Removing it from the default Factory means developers have to add it.  That's fine, but when it's initialized, it needs a bag key passed.  So the service gets initialized with a single key.  I'm guessing when controllers reference the service, they don't pass along their own unique key, and the documentation doesn't mention how to set up the service to accept a key at invocation time.

The session bag documentation could probably use a once-over anyway, as the "Component" example is just a duplicate of the "Controller" example.

**Comment by Alistar84** [comment, marked as answer]
After doing various tests, I added `$this->persistent = new \Phalcon\Session\Bag(get_class());` in the initailize method of my controllers.

This way, it seems to be working as in v3.
